File: app/src/main/java/com/example/waveplayer/media_controller/MediaData.java
package com.example.waveplayer.media_controller;

import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.database.Cursor;
import android.provider.MediaStore;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.room.Room;

import com.example.waveplayer.R;
import com.example.waveplayer.random_playlist.AudioUri;
import com.example.waveplayer.random_playlist.RandomPlaylist;
import com.example.waveplayer.random_playlist.Song;
import com.example.waveplayer.random_playlist.SongDAO;
import com.example.waveplayer.random_playlist.SongDatabase;
import com.example.waveplayer.service_main.ServiceMain;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;

public class MediaData {

    protected static final String SONG_DATABASE_NAME = "SONG_DATABASE_NAME";

    private static final Object MEDIA_DATA_LOCK = new Object();

    private static final String MASTER_PLAYLIST_NAME = "MASTER_PLAYLIST_NAME";

    private static MediaData INSTANCE;

    private final MutableLiveData<String> loadingText = new MutableLiveData<>();

    private final MutableLiveData<Double> loadingProgress = new MutableLiveData<>(0.0);

    private final HashMap<Long, MediaPlayerWUri> songIDToMediaPlayerWUriHashMap = new HashMap<>();

    private final ArrayList<RandomPlaylist> playlists = new ArrayList<>();

    private SongDAO songDAO;

    private RandomPlaylist masterPlaylist;

    private Settings settings =
            new Settings(0.1, 0.1, 0.5, 0);

    /** Returns a singleton instance of this class.
     *  loadData(Context context) must be called for methods on the singleton to function properly.
     * @return A singleton instance of this class that may or may not be loaded with data.
     */
    protected static MediaData getInstance() {
        synchronized (MEDIA_DATA_LOCK) {
            if (INSTANCE == null) {
                INSTANCE = new MediaData();
            }
            return INSTANCE;
        }
    }

    protected LiveData<String> getLoadingText() {
        return loadingText;
    }

    protected LiveData<Double> getLoadingProgress() {
        return loadingProgress;
    }

    protected MediaPlayerWUri getMediaPlayerWUri(Long songID) {
        return songIDToMediaPlayerWUriHashMap.get(songID);
    }

    protected void addMediaPlayerWUri(Long id, MediaPlayerWUri mediaPlayerWURI) {
        songIDToMediaPlayerWUriHashMap.put(id, mediaPlayerWURI);
    }

    protected void releaseMediaPlayers() {
        synchronized (this) {
            Iterator<MediaPlayerWUri> iterator = songIDToMediaPlayerWUriHashMap.values().iterator();
            MediaPlayerWUri mediaPlayerWURI;
            while (iterator.hasNext()) {
                mediaPlayerWURI = iterator.next();
                mediaPlayerWURI.release();
                iterator.remove();
            }
        }
    }

    protected ArrayList<RandomPlaylist> getPlaylists() {
        return playlists;
    }

    protected void addPlaylist(RandomPlaylist randomPlaylist) {
        playlists.add(randomPlaylist);
    }

    protected void addPlaylist(int position, RandomPlaylist randomPlaylist) {
        playlists.add(position, randomPlaylist);
    }

    protected void removePlaylist(RandomPlaylist randomPlaylist) {
        playlists.remove(randomPlaylist);
    }

    protected RandomPlaylist getPlaylist(String playlistName) {
        RandomPlaylist out = null;
        for (RandomPlaylist randomPlaylist : playlists) {
            if (randomPlaylist.getName().equals(playlistName)) {
                out = randomPlaylist;
            }
            break;
        }
        return out;
    }

    protected List<Song> getAllSongs() {
        return masterPlaylist.getSongs();
    }

    protected void setMasterPlaylist(RandomPlaylist masterPlaylist) {
        this.masterPlaylist = masterPlaylist;
    }

    protected RandomPlaylist getMasterPlaylist() {
        return masterPlaylist;
    }

    protected Settings getSettings() {
        return settings;
    }

    protected void setSettings(Settings settings) {
        this.settings = settings;
    }

    protected double getMaxPercent() {
        return settings.maxPercent;
    }

    protected void setMaxPercent(double maxPercent) {
        masterPlaylist.setMaxPercent(maxPercent);
        for (RandomPlaylist randomPlaylist : playlists) {
            randomPlaylist.setMaxPercent(maxPercent);
        }
        settings = new Settings(
                maxPercent, settings.percentChangeUp, settings.percentChangeDown, settings.lowerProb);
    }

    protected void setPercentChangeUp(double percentChangeUp) {
        settings = new Settings(
                settings.maxPercent, percentChangeUp, settings.percentChangeDown, settings.lowerProb);

    }

    protected double getPercentChangeUp() {
        return settings.percentChangeUp;
    }

    protected void setPercentChangeDown(double percentChangeDown) {
        settings = new Settings(
                settings.maxPercent, settings.percentChangeUp, percentChangeDown, settings.lowerProb);
    }

    protected double getPercentChangeDown() {
        return settings.percentChangeDown;
    }

    protected void setLowerProb(double lowerProb) {
        settings = new Settings(
                settings.maxPercent, settings.percentChangeUp, settings.percentChangeDown, lowerProb);
    }

    protected double getLowerProb() {
        return settings.lowerProb;
    }

    protected void loadData(Context context) {
        SongDatabase songDatabase = Room.databaseBuilder(context, SongDatabase.class, SONG_DATABASE_NAME).build();
        songDAO = songDatabase.songDAO();
        SaveFile.loadSaveFile(context, this);
        getSongsFromMediaStore(context);
    }

    private void getSongsFromMediaStore(final Context context) {
        final Resources resources = context.getResources();
        final ArrayList<Song> newSongs = new ArrayList<>();
        final ArrayList<Long> filesThatExist = new ArrayList<>();
        final ExecutorService executorService = ServiceMain.executorServiceFIFO;
        executorService.execute((Runnable) () -> {
            String[] projection = new String[]{
                    MediaStore.Audio.Media._ID, MediaStore.Audio.Media.DISPLAY_NAME,
                    MediaStore.Audio.Media.IS_MUSIC, MediaStore.Audio.Media.ARTIST_ID,
                    MediaStore.Audio.Media.TITLE};
            String selection = MediaStore.Audio.Media.IS_MUSIC + " != ?";
            String[] selectionArgs = new String[]{"0"};
            String sortOrder = MediaStore.Audio.Media.TITLE + " ASC";
            try (Cursor cursor = context.getContentResolver().query(
                    MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                    projection, selection, selectionArgs, sortOrder)) {
                if (cursor != null) {
                    int idCol = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media._ID);
                    int nameCol = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.DISPLAY_NAME);
                    int titleCol = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.TITLE);
                    int artistCol = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ARTIST_ID);
                    final int i = cursor.getCount();
                    int j = 0;
                    loadingText.postValue(resources.getString(R.string.loading1));
                    while (cursor.moveToNext()) {
                        long id = cursor.getLong(idCol);
                        String displayName = cursor.getString(nameCol);
                        String title = cursor.getString(titleCol);
                        String artist = cursor.getString(artistCol);
                        AudioUri audioURI = new AudioUri(displayName, artist, title, id);
                        File file = new File(context.getFilesDir(), String.valueOf(audioURI.id));
                        if (!file.exists() || (songDAO.getSong(id) == null)) {
                            try (FileOutputStream fos = context.openFileOutput(String.valueOf(id), Context.MODE_PRIVATE);
                                 ObjectOutputStream objectOutputStream = new ObjectOutputStream(fos)) {
                                objectOutputStream.writeObject(audioURI);
                            } catch (FileNotFoundException e) {
                                e.printStackTrace();
                            } catch (IOException e) {
                                e.printStackTrace();
                            }
                            newSongs.add(new Song(id, title));
                        }
                        filesThatExist.add(id);
                        loadingProgress.postValue((double) ((double) j) / ((double) i));
                        j++;
                    }
                    loadingText.postValue(resources.getString(R.string.loading2));
                }
            }
        });
        executorService.execute((Runnable) () -> {
            int i = newSongs.size();
            int k = 0;
            for (Song song : newSongs) {
                songDAO.insertAll(song);
                loadingProgress.postValue((double) ((double) k) / ((double) i));
                k++;
            }
        });
        executorService.execute(() -> {
            if (masterPlaylist != null) {
                loadingText.postValue(resources.getString(R.string.loading3));
                addNewSongs(filesThatExist);
                loadingText.postValue(resources.getString(R.string.loading4));
                removeMissingSongs(filesThatExist);
            } else {
                masterPlaylist = new RandomPlaylist(
                        MediaData.MASTER_PLAYLIST_NAME, new ArrayList<>(newSongs),
                        settings.maxPercent, true);
            }
        });
        executorService.execute(() -> {
            if (settings.lowerProb == 0) {
                setLowerProb(2.0 / masterPlaylist.size());
            }
            SaveFile.saveFile(context);
            Intent intent = new Intent();
            intent.addCategory(Intent.CATEGORY_DEFAULT);
            intent.setAction(context.getResources().getString(R.string.broadcast_receiver_action_loaded));
            context.sendBroadcast(intent);
        });
    }

    private void removeMissingSongs(List<Long> filesThatExist) {
        int i = filesThatExist.size();
        int j = 0;
        for (Long songID : masterPlaylist.getSongIDs()) {
            if (!filesThatExist.contains(songID)) {
                masterPlaylist.remove(songDAO.getSong(songID));
                songIDToMediaPlayerWUriHashMap.remove(songID);
                songDAO.delete(songDAO.getSong(songID));
            }
            loadingProgress.postValue((double) ((double) j) / ((double) i));
            j++;
        }
    }

    private void addNewSongs(List<Long> filesThatExist) {
        int i = filesThatExist.size();
        int j = 0;
        for (Long songID : filesThatExist) {
            if (!masterPlaylist.contains(songID)) {
                masterPlaylist.add(songDAO.getSong(songID));
            }
            loadingProgress.postValue((double) ((double) j) / ((double) i));
            j++;
        }
    }

    protected Song getSong(final Long songID) {
        Song song = null;
        try {
            song = ServiceMain.executorServicePool.submit(() -> songDAO.getSong(songID)).get();
        } catch (ExecutionException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return song;
    }

}

File: app/src/main/java/com/example/waveplayer2/media_controller/MediaData.kt
package com.example.waveplayer2.media_controller

import android.content.Context
import android.content.Intent
import android.provider.MediaStore
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.room.Room
import com.example.waveplayer2.R
import com.example.waveplayer2.random_playlist.*
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.ObjectOutputStream
import java.util.*
import java.util.concurrent.Callable
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorService

class MediaData(context: Context) {

    private val _loadingText: MutableLiveData<String> = MutableLiveData()
    val loadingText = _loadingText as LiveData<String>

    private val _loadingProgress: MutableLiveData<Double> = MutableLiveData(0.0)
    val loadingProgress = _loadingProgress as LiveData<Double>

    private val songIDToMediaPlayerWUriHashMap: HashMap<Long, MediaPlayerWUri> = HashMap()
    private val playlists: MutableList<RandomPlaylist> = mutableListOf()
    private var songDAO: SongDAO
    private var masterPlaylist: RandomPlaylist? = null
    private var settings = Settings(0.1, 0.1, 0.5, 0.0)


    fun getMediaPlayerWUri(songID: Long): MediaPlayerWUri? {
        return songIDToMediaPlayerWUriHashMap[songID]
    }

    fun addMediaPlayerWUri(mediaPlayerWURI: MediaPlayerWUri) {
        songIDToMediaPlayerWUriHashMap[mediaPlayerWURI.id] = mediaPlayerWURI
    }

    fun releaseMediaPlayers() {
        // TODO
        synchronized(this) {
            for (mediaPlayerWURI in songIDToMediaPlayerWUriHashMap.values) {
                mediaPlayerWURI.release()
            }
            songIDToMediaPlayerWUriHashMap.clear()
        }
    }

    fun getPlaylists(): List<RandomPlaylist> {
        return playlists
    }

    fun addPlaylist(randomPlaylist: RandomPlaylist) {
        playlists.add(randomPlaylist)
    }

    fun addPlaylist(position: Int, randomPlaylist: RandomPlaylist) {
        playlists.add(position, randomPlaylist)
    }

    fun removePlaylist(randomPlaylist: RandomPlaylist) {
        playlists.remove(randomPlaylist)
    }

    fun getPlaylist(playlistName: String): RandomPlaylist? {
        var out: RandomPlaylist? = null
        for (randomPlaylist in playlists) {
            if (randomPlaylist.getName() == playlistName) {
                out = randomPlaylist
            }
            break
        }
        return out
    }

    fun getAllSongs(): List<Song>? {
        return masterPlaylist?.getSongs()
    }

    fun setMasterPlaylist(masterPlaylist: RandomPlaylist) {
        this.masterPlaylist = masterPlaylist
    }

    fun getMasterPlaylist(): RandomPlaylist? {
        return masterPlaylist
    }

    fun getSettings(): Settings {
        return settings
    }

    fun setSettings(settings: Settings) {
        this.settings = settings
    }

    fun getMaxPercent(): Double {
        return settings.maxPercent
    }

    fun setMaxPercent(maxPercent: Double) {
        masterPlaylist?.setMaxPercent(maxPercent)
        for (randomPlaylist in playlists) {
            randomPlaylist.setMaxPercent(maxPercent)
        }
        settings = Settings(
                maxPercent, settings.percentChangeUp, settings.percentChangeDown, settings.lowerProb)
    }

    fun setPercentChangeUp(percentChangeUp: Double) {
        settings = Settings(
                settings.maxPercent, percentChangeUp, settings.percentChangeDown, settings.lowerProb)
    }

    fun getPercentChangeUp(): Double {
        return settings.percentChangeUp
    }

    fun setPercentChangeDown(percentChangeDown: Double) {
        settings = Settings(
                settings.maxPercent, settings.percentChangeUp, percentChangeDown, settings.lowerProb)
    }

    fun getPercentChangeDown(): Double {
        return settings.percentChangeDown
    }

    fun setLowerProb(lowerProb: Double) {
        settings = Settings(
                settings.maxPercent, settings.percentChangeUp, settings.percentChangeDown, lowerProb)
    }

    fun getLowerProb(): Double {
        return settings.lowerProb
    }

    fun loadData(context: Context) {
         SaveFile.loadSaveFile(context, this)
         getSongsFromMediaStore(context)
    }

    private fun getSongsFromMediaStore(context: Context) {
        val resources = context.resources
        val newSongs = mutableListOf<Song>()
        val filesThatExist = mutableListOf<Long>()
        val executorServiceFIFO: ExecutorService = ServiceMain.executorServiceFIFO
        executorServiceFIFO.execute {
            val projection = arrayOf(
                    MediaStore.Audio.Media._ID, MediaStore.Audio.Media.DISPLAY_NAME,
                    MediaStore.Audio.Media.IS_MUSIC, MediaStore.Audio.Media.ARTIST_ID,
                    MediaStore.Audio.Media.TITLE)
            val selection = MediaStore.Audio.Media.IS_MUSIC + " != ?"
            val selectionArgs = arrayOf("0")
            val sortOrder = MediaStore.Audio.Media.TITLE + " ASC"
            context.contentResolver.query(
                    MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                    projection, selection, selectionArgs, sortOrder).use { cursor ->
                if (cursor != null) {
                    val idCol = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media._ID)
                    val nameCol = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.DISPLAY_NAME)
                    val titleCol = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.TITLE)
                    val artistCol = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ARTIST_ID)
                    val i = cursor.count
                    var j = 0
                    _loadingText.postValue(resources.getString(R.string.loading1))
                    while (cursor.moveToNext()) {
                        val id = cursor.getLong(idCol)
                        val displayName = cursor.getString(nameCol)
                        val title = cursor.getString(titleCol)
                        val artist = cursor.getString(artistCol)
                        val file = File(context.filesDir, id.toString())
                        // TODO put in user triggered scan
                        // !file.exists() || (songDAO.getSong(id) == null)
                        if ((masterPlaylist?.contains(Song(id, title)) != true)) {
                            val audioURI = AudioUri(displayName, artist, title, id)
                            try {
                                context.openFileOutput(id.toString(), Context.MODE_PRIVATE).use { fos ->
                                    ObjectOutputStream(fos).use { objectOutputStream ->
                                        objectOutputStream.writeObject(audioURI)
                                    }
                                }
                            } catch (e: FileNotFoundException) {
                                e.printStackTrace()
                            } catch (e: IOException) {
                                e.printStackTrace()
                            }
                            newSongs.add(Song(id, title))
                        }
                        filesThatExist.add(id)
                        _loadingProgress.postValue(j.toDouble() / i.toDouble())
                        j++
                    }
                    _loadingText.postValue(resources.getString(R.string.loading2))
                }
            }
        }
        executorServiceFIFO.execute {
            val i = newSongs.size
            _loadingText.postValue(context.getResources().getString(R.string.loading2))
            for ((k, song) in newSongs.withIndex()) {
                songDAO.insertAll(song)
                _loadingProgress.postValue(k.toDouble() / i.toDouble())
            }
        }
        executorServiceFIFO.execute {
            if (masterPlaylist != null) {
                _loadingText.postValue(resources.getString(R.string.loading3))
                addNewSongs(filesThatExist)
                _loadingText.postValue(resources.getString(R.string.loading4))
                removeMissingSongs(filesThatExist)
            } else {
                masterPlaylist = RandomPlaylist(
                        MASTER_PLAYLIST_NAME,
                        ArrayList(newSongs),
                        settings.maxPercent,
                        true
                )
            }
        }
        executorServiceFIFO.execute {
            if (settings.lowerProb == 0.0) {
                setLowerProb(2.0 / (masterPlaylist?.size()?:2).toDouble())
            }
            SaveFile.saveFile(context)
            val intent = Intent()
            intent.addCategory(Intent.CATEGORY_DEFAULT)
            intent.action = context.resources.getString(R.string.broadcast_receiver_action_loaded)
            context.sendBroadcast(intent)
        }
    }

    private fun removeMissingSongs(filesThatExist: List<Long>) {
        val i = filesThatExist.size
        val ids = masterPlaylist?.getSongIDs()
        if(ids != null) {
            for((j, songID) in ids.withIndex()) {
                if (!filesThatExist.contains(songID)) {
                    songDAO.getSong(songID)?.let {
                        masterPlaylist?.remove(it)
                        songDAO.delete(it) }
                    songIDToMediaPlayerWUriHashMap.remove(songID)
                }
                _loadingProgress.postValue(j.toDouble() / i.toDouble())
            }
        }
    }

    private fun addNewSongs(filesThatExist: List<Long>) {
        val i = filesThatExist.size
        for ((j, songID) in filesThatExist.withIndex()) {
            if (masterPlaylist?.contains(songID) == false) {
                songDAO.getSong(songID)?.let { masterPlaylist?.add(it) }
            }
            _loadingProgress.postValue(j.toDouble() / i.toDouble())
        }
    }

    fun getSong(songID: Long): Song? {
        var song: Song? = null
        try {
            song = ServiceMain.executorServicePool.submit(Callable { songDAO.getSong(songID) }).get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        return song
    }

    init {
        val songDatabase = Room.databaseBuilder(context, SongDatabase::class.java, SONG_DATABASE_NAME).build()
        songDAO = songDatabase.songDAO()
    }

    companion object {
        const val SONG_DATABASE_NAME: String = "SONG_DATABASE_NAME"
        private val MEDIA_DATA_LOCK: Any = Any()
        private const val MASTER_PLAYLIST_NAME: String = "MASTER_PLAYLIST_NAME"
        private var INSTANCE: MediaData? = null

        /** Returns a singleton instance of this class.
         * loadData(Context context) must be called for methods on the singleton to function properly.
         * @return A singleton instance of this class that may or may not be loaded with data.
         */
        fun getInstance(context: Context): MediaData {
            synchronized(MEDIA_DATA_LOCK) {
                if (INSTANCE == null) {
                    INSTANCE = MediaData(context)
                }
                return INSTANCE!!
            }
        }
    }
}